src/imitation/algorithms/bc.py

```diff
@@ -4,178 +4,261 @@
-                end of every batch.
-            progress_bar_visible: If True, then show a tqdm progress bar.
+    # TODO(shwang): Maybe instead use a callback that can be shared between
+    #   all algorithms' `.train()` for generating rollout stats.
+    #   EvalCallback could be a good fit:
+    #   https://stable-baselines3.readthedocs.io/en/master/guide/callbacks.html#evalcallback
 
-        Raises:
-            ValueError: If neither or both of `n_epochs` and `n_batches` are non-None.
-        """
-        if n_epochs is not None and n_batches is None:
-            self.use_epochs = True
-        elif n_epochs is None and n_batches is not None:
-            self.use_epochs = False
-        else:
-            raise ValueError(
-                "Must provide exactly one of `n_epochs` and `n_batches` arguments.",
+    def __call__(self, policy: policies.ActorCriticPolicy) -> Mapping[str, float]:
+        if self.venv is not None and self.n_episodes > 0:
+            trajs = rollout.generate_trajectories(
+                policy,
+                self.venv,
+                rollout.make_min_episodes(self.n_episodes),
             )
+            return rollout.rollout_stats(trajs)
+        else:
+            return dict()
 
-        self.data_loader = data_loader
-        self.n_epochs = n_epochs
-        self.n_batches = n_batches
-        self.on_epoch_end = on_epoch_end
-        self.on_batch_end = on_batch_end
-        self.progress_bar_visible = progress_bar_visible
 
-    def __iter__(
+class BCLogger:
+    """Utility class to help logging information relevant to Behavior Cloning."""
+
+    def __init__(self, logger: imit_logger.HierarchicalLogger):
+        """Create new BC logger.
+
+        Args:
+            logger: The logger to which to feed all the information.
+        """
+        self._logger = logger
+        self._tensorboard_step = 0
+
+    def reset_tensorboard_steps(self):
+        self._tensorboard_step = 0
+
+    def log_epoch(self, epoch_number):
+        self._logger.record("bc/epoch", epoch_number)
+
+    def log_batch(
         self,
-    ) -> Iterable[Tuple[algo_base.TransitionMapping, Mapping[str, Any]]]:
-        """Yields batches while updating tqdm display to display progress."""
-        samples_so_far = 0
-        epoch_num = 0
-        batch_num = 0
-        batch_suffix = epoch_suffix = ""
-        if self.progress_bar_visible:
-            if self.use_epochs:
-                display = tqdm.tqdm(total=self.n_epochs)
-                epoch_suffix = f"/{self.n_epochs}"
-            else:  # Use batches.
-                display = tqdm.tqdm(total=self.n_batches)
-                batch_suffix = f"/{self.n_batches}"
-        else:
-            display = _NoopTqdm()
+        batch_num: int,
+        batch_size: int,
+        num_samples_so_far: int,
+        loss: BehaviorCloningLoss,
+        rollout_stats: Mapping[str, float],
+    ):
+        self._tensorboard_step += 1
```

Comment:
Feels a bit more natural to me to increment at end than at beginning (so we're 0-indexed), but doesn't really matter.

Reply:
True. No idea how it ended up there. I moved it down.